simpl/incubator/rest.py

```diff
@@ -190,3 +154,41 @@ def wrapped(*args, **kwargs):
                 bottle.abort(400, str(exc))
         return wrapped
     return deco
+
+
+def body(schema=None, types=None, required=False, default=None):
+    """Decorator to parse and validate API body.
+
+    :keyword schema: callable that accepts raw data and returns the coerced (or
+        unchanged) data if it is valid. It should raise an error if the data is
+        not valid.
+    :keyword types: supported content types (default is ['application/json'])
+    :keyword required: if true and no body specified will raise an error.
+    :keyword default: default value to return if no body supplied
+
+    Note: only json types are supported.
+    """
+    if not types:
+        types = ['application/json']
+    if not all('json' in t for t in types):
+        raise NotImplementedError("Only 'json' body supported.")
+
+    def wrap(fxn):
+        """Return a decorated callable."""
+        def wrapped(*args, **kwargs):
+            """Callable to called when the decorated function is called."""
+            body = bottle.request.json
+            if required and not body:
+                bottle.abort(400, "Call body cannot be empty")
+            if body is None:
+                body = default
+            if schema:
+                try:
+                    body = schema(body)
+                except (KeyboardInterrupt, SystemExit):
+                    raise  # don't catch and ignore attempts to end the app
+                except Exception as exc:
+                    bottle.abort(400, str(exc))
+            return fxn(*args, body=body, **kwargs)
```

Comment:
As I type up a trivial example below to test that it achieves:
- bottle still directly accessible
- works with other `simpl.rest` decorators 
- readable, intuitive, easy to use

It feels fine for POST:
```python
import bottle

from simpl import rest

@bottle.post('/v1/widgets/<api_id>')
@rest.query(schema={'force': bool}, defaults={'force': False})
@rest.body(schema=WIDGET)
def edit_widget(body, api_id, query=None):
    """Edit an existing widget."""
    assert query and 'force' in query, "Never, ever fails! Of course..."

    # If I want to accommodate those that say query params in POSTs are bad,
    # I can still access bottle directly.
    if 'X-Force' in bottle.request.headers:
        query['force'] = bottle.request.headers['X-Force']

    try:
        save(api_id, body, force=query['force'])
    except IndexError:
        raise rest.HTTPError("No widget with id '%s'" % api_id, http_code=404)
```

I came across issues with GET which I'll post shortly in the code.